File: 26-Visitor/src/go/print_visitor.go
package main

import "fmt"

// PrintVisitor renders the expression as infix text. Adding this operation
// required NO change to any element class.
type PrintVisitor struct {
	result string
}

func (p *PrintVisitor) Result() string { return p.result }

func (p *PrintVisitor) VisitNumber(number *NumberExpr) {
	p.result = fmt.Sprintf("%g", number.Value())
}

func (p *PrintVisitor) VisitAddition(addition *Addition) {
	addition.Left().Accept(p)
	left := p.result
	addition.Right().Accept(p)
	right := p.result
	p.result = "(" + left + " + " + right + ")"
}

func (p *PrintVisitor) VisitMultiplication(multiplication *Multiplication) {
	multiplication.Left().Accept(p)
	left := p.result
	multiplication.Right().Accept(p)
	right := p.result
	p.result = "(" + left + " * " + right + ")"
}

File: 26-Visitor/src/rust/print_visitor.rs
use crate::visitor::Visitor;
use crate::number_expr::NumberExpr;
use crate::addition::Addition;
use crate::multiplication::Multiplication;

/// Concrete Visitor — renders the expression as infix text. Adding this
/// operation required NO change to any element type.
pub struct PrintVisitor {
    pub result: String,
}

impl PrintVisitor {
    pub fn new() -> Self {
        PrintVisitor { result: String::new() }
    }
}

impl Visitor for PrintVisitor {
    fn visit_number(&mut self, number: &NumberExpr) {
        self.result = format!("{}", number.value());
    }

    fn visit_addition(&mut self, addition: &Addition) {
        addition.left().accept(self);
        let left = self.result.clone();
        addition.right().accept(self);
        let right = self.result.clone();
        self.result = format!("({} + {})", left, right);
    }

    fn visit_multiplication(&mut self, multiplication: &Multiplication) {
        multiplication.left().accept(self);
        let left = self.result.clone();
        multiplication.right().accept(self);
        let right = self.result.clone();
        self.result = format!("({} * {})", left, right);
    }
}
